## Computer Assignment #1 Report
## Artificial Intelligence Course - Spring 1400 - University of Tehran
## Daneshvar Amrollahi (Student ID: 810197685)

### Each state (node) of the search problem is represented with a tuple $((x, y), backpack, balls\_done)$ where
#### $(x, y)$ denotes the current location in the grid
#### $backpack$ is a binary string of the length of number of balls. $backpack[i] = 1 \Leftrightarrow$ ball number $i$ is in our backpack
#### $balls\_done$ is a binary string of the length of number of balls. $backpack = 1 \Leftrightarrow$ ball number $i$ is put in it's final location

In [450]:
f = open("1.txt", "r")

lines = []
for line in f:
    l = line.rstrip("\n")
    lines.append(l)
    
print(lines)


n, m = map(int, lines[0].split())

start_x, start_y = map(int, list(lines[1].split()))

finish_x, finish_y = map(int, lines[2].split())

c = int(lines[3])

k = int(lines[4])

ball = []
for i in range(k):
    x1, y1, x2, y2 = map(int, lines[5 + i].split())
    ball.append((x1, y1, x2, y2))

    
grid = []
for i in range(n):
    row = ""
    for ch in lines[5 + k + i]:
        if ch != ' ':
            row += ch
    grid.append(row)

['5 7', '1 0', '2 6', '1', '3', '1 2 2 5', '2 3 3 3', '3 2 1 1 ', '* * * * * * *', '- - - - - - *', '* * * - - - -', '* - - - - - *', '* * * * * * *']


In [451]:
ball

[(1, 2, 2, 5), (2, 3, 3, 3), (3, 2, 1, 1)]

In [452]:
grid

['*******', '------*', '***----', '*-----*', '*******']

In [453]:
c

1

In [454]:
start_state = ( (start_x, start_y), '0' * len(ball), '0' * len(ball) ) #loc, backpack, balls_done

### According to the model explained above, the $start\_state$ would be $( (start\_x, start\_y), 00..0, 00..0 )$
### In the beginning no balls are in our backpack and also no balls are put in their desired location. Therefore both $backpack$ and $balls\_done$ would be 00..0

In [455]:
start_state

((1, 0), '000', '000')

In [456]:
goal_state =  ( (finish_x, finish_y), '0' * len(ball), '1' * len(ball) )

### According to the model explained above, the $goal\_state$ would be $( (finish\_x, finish\_y), 00..0, 11..1 )$
### In the ending no balls are in our backpack and also all balls are put in their desired location. Therefore $backpack = 00..0$ and $balls\_done = 11..1$

In [457]:
goal_state

((2, 6), '000', '111')

## BFS (Breadth First Search)

In [458]:
import copy
import time



def print_path(cur):
    print(cur)
    if (cur == start_state):
          return
    print_path(par[cur])


start_state = ( (start_x, start_y), '0' * len(ball), '0' * len(ball) ) #loc, backpack, balls_done
goal_state =  ( (finish_x, finish_y), '0' * len(ball), '1' * len(ball) )

mark = dict()
mark[start_state] = 1
q = []
q.append(start_state)
dist = dict()
dist[start_state] = 0

par = dict()

while (len(q) > 0):
    cur = q.pop(0)
    
    if (cur == goal_state):
        break
    
    x, y = cur[0][0], cur[0][1]
    backpack = cur[1]
    balls_done = cur[2]
    
    
    #Put down a ball in no moves
    for i in range(len(backpack)):
        if (backpack[i] == '0'): #dont have it in our backpack
            continue
        if (ball[i][2] == x and ball[i][3] == y):
            backpack = backpack[:i] + "0" + backpack[i + 1:]
            balls_done = balls_done[:i] + "1" + balls_done[i + 1:]
                    
    
    #print(x, y, backpack, balls_done, dist[cur])
    
    
    #Pick up a ball
    for i in range(len(ball)):
        if (balls_done[i] == '1'): #Ball is already in final destination 
            continue
        if (backpack[i] == '1'): #Ball is already in backpack
            continue
            
        ballx, bally = ball[i][0], ball[i][1]
        if (ballx == x and bally == y):
            if (sum( map(int, list(backpack))) < c):
                new_backpack = copy.deepcopy(backpack)
                new_backpack = new_backpack[:i] + "1" + new_backpack[i + 1:] 
                
                new_state = ((x, y), new_backpack, balls_done)
                    
                if (not new_state in mark):
                    mark[new_state] = 1
                    q.append(new_state)
                    dist[new_state] = dist[cur] + 1
                    par[new_state] = cur
                
    
    #Go down
    if (x + 1 < n):
        if (grid[x + 1][y] != '*'):
            new_state = ((x + 1, y), backpack, balls_done)
            
            if (not new_state in mark):
                mark[new_state] = 1
                q.append(new_state)
                dist[new_state] = dist[cur] + 1
                par[new_state] = cur
    
    #Go up
    if (x - 1 >= 0):
        if (grid[x - 1][y] != '*'):
            new_state = ((x - 1, y), backpack, balls_done)
            
            if (not new_state in mark):
                mark[new_state] = 1
                q.append(new_state)
                dist[new_state] = dist[cur] + 1
                par[new_state] = cur
    
    #Go right
    if (y + 1 < m):
        if (grid[x][y + 1] != '*'):
            new_state = ((x, y + 1), backpack, balls_done)
                
            if (not new_state in mark):
                mark[new_state] = 1
                q.append(new_state)
                dist[new_state] = dist[cur] + 1
                par[new_state] = cur
    #Go left
    if (y - 1 >= 0):
        if (grid[x][y - 1] != '*'):
            new_state = ((x, y - 1), backpack, balls_done)
            
            if (not new_state in mark):
                mark[new_state] = 1
                q.append(new_state)
                dist[new_state] = dist[cur] + 1
                par[new_state] = cur
                

st = time.time()
print("The path len is {}".format(dist[goal_state]))
fn = time.time()
print("The path is:")
print_path(goal_state)
print("Number of states seen", len(dist))
print("Time consumed is %s seconds" %(fn - st))

The path len is 20
The path is:
((2, 6), '000', '111')
((2, 5), '100', '011')
((2, 4), '100', '011')
((2, 3), '100', '011')
((1, 3), '100', '011')
((1, 2), '100', '011')
((1, 2), '000', '011')
((1, 1), '001', '010')
((1, 2), '001', '010')
((1, 3), '001', '010')
((2, 3), '001', '010')
((3, 3), '001', '010')
((3, 2), '001', '010')
((3, 2), '000', '010')
((3, 3), '010', '000')
((2, 3), '010', '000')
((2, 3), '000', '000')
((1, 3), '000', '000')
((1, 2), '000', '000')
((1, 1), '000', '000')
((1, 0), '000', '000')
Number of states seen 267
Time consumed is 0.00037169456481933594 seconds


### Results of BFS:

| --- | Distance with Goal | Satates Seen | Unique States Seen | Avg Exec Time |
| --- | --- | --- | --- | --- |
| Test 1 | 20 | 267 | 267 | 0.0001s |
| Test 2 | 48 | 2304 | 2304 | 0.0002s |
| Test 3 | 68 | 221589 | 221589 | 0.0001s |
| Test 4 | 92 | 457 | 457 | 0.0001s |

The number of states seen is equal to uniqute states seen since this algorithm doesn't explore a state more than once.

### Advantages:
1. The solution will definitely found out by BFS If there is some solution.
2. BFS will never get trapped in a blind alley, which means unwanted nodes.
3. If there is more than one solution then it will find a solution with minimal steps.

### Disadvantages:
1. Memory Constraints As it stores all the nodes of the present level to go for the next level.
2. If a solution is far away then it consumes time.

In [459]:
mark.clear()
par.clear()
dist.clear()

In [460]:
start_state

((1, 0), '000', '000')

In [461]:
goal_state

((2, 6), '000', '111')

## IDS (Iterative Deepening Search)

In [462]:
import copy
import time


def print_path(cur):
    print(cur)
    if (cur == start_state):
          return
    print_path(par[cur])

seen = dict()

start_state = ( (start_x, start_y), '0' * len(ball), '0' * len(ball) ) #loc, backpack, balls_done
goal_state =  ( (finish_x, finish_y), '0' * len(ball), '1' * len(ball) )
dist = dict()
par = dict()

total_states = 0

def DLS(v, p, target, depth, limit):
    if (depth == limit):
        return
    global total_states
    total_states += 1
    par[v] = p
    if (v != p):
        dist[v] = dist[p] + 1
    seen[v] = 1
    if (v == target):
        return True
    
    
    x, y = v[0][0], v[0][1]
    backpack = v[1]
    balls_done = v[2]
    
    
    #Put down a ball in no moves
    for i in range(len(backpack)):
        if (backpack[i] == '0'): #dont have it in our backpack
            continue
        if (ball[i][2] == x and ball[i][3] == y):
            backpack = backpack[:i] + "0" + backpack[i + 1:]
            balls_done = balls_done[:i] + "1" + balls_done[i + 1:]
                    
    
    #nv = ((x, y), backpack, balls_done)
    #print(nv)
    #print(dist)
    #print(par)
    #print(seen)
    #print()
    
    
    #Pick up a ball
    for i in range(len(ball)):
        if (balls_done[i] == '1'): #Ball is already in final destination 
            continue
        if (backpack[i] == '1'): #Ball is already in backpack
            continue
            
        ballx, bally = ball[i][0], ball[i][1]
        if (ballx == x and bally == y):
            if (sum( map(int, list(backpack))) < c):
                new_backpack = copy.deepcopy(backpack)
                new_backpack = new_backpack[:i] + "1" + new_backpack[i + 1:] 
                
                new_state = ((x, y), new_backpack, balls_done)
                if (not new_state in seen):
                    if (DLS(new_state, v, target, depth + 1, limit)):
                        #del seen[v]
                        return True
                elif dist[new_state] > dist[v] + 1:
                    if (DLS(new_state, v, target, depth + 1, limit)):
                        #del seen[v]
                        return True
                
                
    
    #Go down
    if (x + 1 < n):
        if (grid[x + 1][y] != '*'):
            new_state = ((x + 1, y), backpack, balls_done)
            if (not new_state in seen):
                if (DLS(new_state, v, target, depth + 1, limit)):
                    del seen[v]
                    return True
            elif dist[new_state] > dist[v] + 1:
                    if (DLS(new_state, v, target, depth + 1, limit)):
                        #del seen[v]
                        return True
            
    
    #Go up
    if (x - 1 >= 0):
        if (grid[x - 1][y] != '*'):
            new_state = ((x - 1, y), backpack, balls_done)
            if (not new_state in seen):
                if (DLS(new_state, v, target, depth + 1, limit)):
                    del seen[v]
                    return True
            elif dist[new_state] > dist[v] + 1:
                    if (DLS(new_state, v, target, depth + 1, limit)):
                        #del seen[v]
                        return True
            
    
    #Go right
    if (y + 1 < m):
        if (grid[x][y + 1] != '*'):
            new_state = ((x, y + 1), backpack, balls_done)
            if (not new_state in seen):
                if (DLS(new_state, v, target, depth + 1, limit)):
                    del seen[v]
                    return True 
            elif dist[new_state] > dist[v] + 1:
                    if (DLS(new_state, v, target, depth + 1, limit)):
                        #del seen[v]
                        return True
            
    #Go left
    if (y - 1 >= 0):
        if (grid[x][y - 1] != '*'):
            new_state = ((x, y - 1), backpack, balls_done)
            if (not new_state in seen):
                if (DLS(new_state, v, target, depth + 1, limit)):
                    del seen[v]
                    return True   
            elif dist[new_state] > dist[v] + 1:
                    if (DLS(new_state, v, target, depth + 1, limit)):
                        #del seen[v]
                        return True
    
    #del seen[v]
    return False

def IDDFS(init, goal, max_depth):
    dist[init] = 0
    for i in range(max_depth + 1):
        print("Checking depth ", i)
        seen.clear()
        res = DLS(init, init, goal, 0, i) 
        if (res == True):
            return True
    return False

st = time.time()
res = IDDFS(start_state, goal_state, 100)
fn = time.time()
if (res):
    print("Path found with IDS")
    print("Len of path is ", dist[goal_state])
    print_path(goal_state)
    print("Time consumed is %s seconds" %(fn - st))
    print("Unique states seen", len(par))
    print("Total states seen", total_states)
    
else:
    print("Path not found with IDS")

Checking depth  0
Checking depth  1
Checking depth  2
Checking depth  3
Checking depth  4
Checking depth  5
Checking depth  6
Checking depth  7
Checking depth  8
Checking depth  9
Checking depth  10
Checking depth  11
Checking depth  12
Checking depth  13
Checking depth  14
Checking depth  15
Checking depth  16
Checking depth  17
Checking depth  18
Checking depth  19
Checking depth  20
Checking depth  21
Path found with IDS
Len of path is  20
((2, 6), '000', '111')
((2, 5), '100', '011')
((2, 4), '100', '011')
((2, 3), '100', '011')
((1, 3), '100', '011')
((1, 2), '100', '011')
((1, 2), '000', '011')
((1, 1), '001', '010')
((1, 2), '001', '010')
((1, 3), '001', '010')
((2, 3), '001', '010')
((3, 3), '001', '010')
((3, 2), '001', '010')
((3, 2), '000', '010')
((3, 3), '010', '000')
((2, 3), '010', '000')
((2, 3), '000', '000')
((1, 3), '000', '000')
((1, 2), '000', '000')
((1, 1), '000', '000')
((1, 0), '000', '000')
Time consumed is 0.03508639335632324 seconds
Unique states seen 257
To

### Results of IDS:

| --- | Distance with Goal | Satates Seen | Unique States Seen | Avg Exec Time |
| --- | --- | --- | --- | --- |
| Test 1 | 20 | 3704 | 257 | 0.18s |
| Test 2 | 48 | 405819 | 2304 | 1.61s |
| Test 3 | 68 | 207353 | 7635544 | 66s |
| Test 4 | 92 | 457 | 128832 | 0.59s |

The number of states seen is equal to uniqute states seen since this algorithm doesn't explore a state more than once.

### Advantages
1. When the solutions are found at the lower depths say n, then the algorithm proves to be efficient and in time.

### Disadvantages
1. The main problem with IDDFS is the time and wasted calculations that take place at each depth.

# A* 

### The first heuristic we use is the manhattan distance: $manhattan(x_1, y_1, x_2, y_2) = |x_1 - x_2| + |y_1 - y_2|$

$h(node) = manhattan(node.x, node.y, x\_goal, y\_goal)$

### In each step the node with minimum $g(node) + h(node)$ is chosen where $g(node)$ is the cost of reaching this node from init state. 

#### In the implementation below, the parameter $g$ is not added to the state itself. It is added alonside each to state t $openset$ and $closedset$ like $(g, state)$

#### For example if we have a state $((2, 3), 010, 100)$ with cost $g = 11$ until now, the tuple $(11,((2, 3), 010, 100))$ is added to $openset$ and $closedset$

In [463]:
from _heapq import heappush, heappop


import time

start_state = ( (start_x, start_y), '0' * len(ball), '0' * len(ball) ) #loc, backpack, balls_done
goal_state =  ( (finish_x, finish_y), '0' * len(ball), '1' * len(ball) )
par = dict()

def print_path(cur):
    print(cur)
    if (cur == start_state):
          return
    print_path(par[cur])

def manhattan(x1, y1, x2, y2):
    return abs(x1 - x2) + abs(y1 - y2)
    
def get_heuristic(node, goal):
    return manhattan(node[0][0], node[0][1], goal[0][0], goal[0][1])

    
is_in_heap = dict()
    
    
def aStar(start, goal):
    openset = []
    closedset = set([])
    explored = 0
    current = start
    
    print("Start is ", start)
    
    #openset.append((0, current))
    heappush(openset, (0 + get_heuristic(current, goal), current))
    is_in_heap[(0 + get_heuristic(current, goal), current)] = 1
    
    
    while openset:
        explored += 1
        current = heappop(openset)
        del is_in_heap[current]
            
        if current[1] == goal:    
            plen = 0
            node = current[1]
            while (node != start):
                node = par[node]
                plen += 1
            print("Path found with len", plen)
            print_path(goal_state)
            return explored
        
        closedset.add(current[1])
        
        x, y = current[1][0][0], current[1][0][1]
        backpack = current[1][1]
        balls_done = current[1][2]
        cur_gh = current[0]

        
        
        #Put down a ball in no moves
        for i in range(len(backpack)):
            if (backpack[i] == '0'): #dont have it in our backpack
                continue
            if (ball[i][2] == x and ball[i][3] == y):
                backpack = backpack[:i] + "0" + backpack[i + 1:]
                balls_done = balls_done[:i] + "1" + balls_done[i + 1:]
                
                closedset.add(((x, y), backpack, balls_done))
        
        for i in range(len(ball)):
            if (balls_done[i] == '1'): #Ball is already in final destination 
                continue
            if (backpack[i] == '1'): #Ball is already in backpack
                continue
            
            ballx, bally = ball[i][0], ball[i][1] #balls initial location
            if (ballx == x and bally == y):
                if (sum( map(int, list(backpack))) < c): #have enough capacity
                    new_backpack = copy.deepcopy(backpack)
                    new_backpack = new_backpack[:i] + "1" + new_backpack[i + 1:] 
                    
                    
                    new_state = ((x, y), new_backpack, balls_done)
                    
                    if (not new_state in closedset) and (not (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state) in is_in_heap):
                        par[new_state] = current[1]
                        heappush(openset, (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state))
                        is_in_heap[(cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state)] = 1
                        
        
        if (x + 1 < n):
            if (grid[x + 1][y] != '*'):
                
                new_state = ((x + 1, y), backpack, balls_done)
                if (not new_state in closedset) and (not (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state) in is_in_heap):
                    par[new_state] = current[1]
                    heappush(openset, (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state))
                    is_in_heap[(cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state)] = 1
                    
        if (x - 1 >= 0):
            if (grid[x - 1][y] != '*'):
                new_state = ((x - 1, y), backpack, balls_done)
                
                if (not new_state in closedset) and (not (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state) in is_in_heap):
                    par[new_state] = current[1]
                    heappush(openset, (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state))
                    is_in_heap[(cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state)] = 1
       
        
        if (y + 1 < m):
            if (grid[x][y + 1] != '*'):
                new_state = ((x, y + 1), backpack, balls_done)
                
                if (not new_state in closedset) and (not (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state) in is_in_heap):
                    par[new_state] = current[1]
                    heappush(openset, (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state))
                    is_in_heap[(cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state)] = 1
        
        if (y - 1 >= 0):
            if (grid[x][y - 1] != '*'):
                new_state = ((x, y - 1), backpack, balls_done)
                if (not new_state in closedset) and (not (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state) in is_in_heap):
                    par[new_state] = current[1]
                    heappush(openset, (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state))
                    is_in_heap[(cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state)] = 1
        
st = time.time()
explored = aStar(start_state, goal_state)
fn = time.time()
print("Number of states explored is", explored)
print("Time consumed is %s" %(fn - st))


Start is  ((1, 0), '000', '000')
Path found with len 20
((2, 6), '000', '111')
((2, 5), '100', '011')
((1, 5), '100', '011')
((1, 4), '100', '011')
((1, 3), '100', '011')
((1, 2), '100', '011')
((1, 2), '000', '011')
((1, 1), '001', '010')
((1, 2), '001', '010')
((1, 3), '001', '010')
((2, 3), '001', '010')
((3, 3), '001', '010')
((3, 2), '001', '010')
((3, 2), '000', '010')
((3, 3), '010', '000')
((2, 3), '010', '000')
((2, 3), '000', '000')
((1, 3), '000', '000')
((1, 2), '000', '000')
((1, 1), '000', '000')
((1, 0), '000', '000')
Number of states explored is 205
Time consumed is 0.0037887096405029297


### Results of A* with Manhattan heuristic:

| --- | Distance with Goal | Satates Seen | Unique States Seen | Avg Exec Time |
| --- | --- | --- | --- | --- |
| Test 1 | 20 | 205 | 205 | 0.004s |
| Test 2 | 48 | 625 | 625 | 0.008s |
| Test 3 | 68 | 98885 | 98885 | 1.87 |
| Test 4 | 92 | 457 | 457 | 0.012s |

The number of states seen is equal to uniqute states seen since this algorithm doesn't explore a state more than once.

### The next heuristic we will be using is:
#### $h(node) = Euclidean(node.x, node.y, goal.x, goal.y)$ where
$ Euclidean(x1, y1, x2, y2) = \sqrt{(x_1 - x_2)^2 + (y_1 - y_2)^2}$ 

In [464]:
from _heapq import heappush, heappop
from math import sqrt

import time

start_state = ( (start_x, start_y), '0' * len(ball), '0' * len(ball) ) #loc, backpack, balls_done
goal_state =  ( (finish_x, finish_y), '0' * len(ball), '1' * len(ball) )
par = dict()

def print_path(cur):
    print(cur)
    if (cur == start_state):
          return
    print_path(par[cur])

def manhattan(x1, y1, x2, y2):
    return abs(x1 - x2) + abs(y1 - y2)
    
def euclidean(x1, y1, x2, y2):
    return sqrt( (x1 - x2)**2 + (y1 - y2)**2 )
    
def get_heuristic(node, goal):
    return euclidean(node[0][0], node[0][1], goal[0][0], goal[0][1]) 
    
is_in_heap = dict()
    


    
def aStar(start, goal):
    openset = []
    closedset = set([])
    explored = 0
    current = start
    
    print("Start is ", start)
    
    #openset.append((0, current))
    heappush(openset, (0 + get_heuristic(current, goal), current))
    is_in_heap[(0 + get_heuristic(current, goal), current)] = 1
    
    
    while openset:
    
        current = heappop(openset)
        explored += 1
        del is_in_heap[current]
            
        if current[1] == goal:    
            plen = 0
            node = current[1]
            while (node != start):
                node = par[node]
                plen += 1
            print("Path found with len", plen)
            print_path(goal_state)
            return explored
        
        closedset.add(current[1])
        
        x, y = current[1][0][0], current[1][0][1]
        backpack = current[1][1]
        balls_done = current[1][2]
        cur_gh = current[0]

        
        
        #Put down a ball in no moves
        for i in range(len(backpack)):
            if (backpack[i] == '0'): #dont have it in our backpack
                continue
            if (ball[i][2] == x and ball[i][3] == y):
                backpack = backpack[:i] + "0" + backpack[i + 1:]
                balls_done = balls_done[:i] + "1" + balls_done[i + 1:]
                
                closedset.add(((x, y), backpack, balls_done))
        
        for i in range(len(ball)):
            if (balls_done[i] == '1'): #Ball is already in final destination 
                continue
            if (backpack[i] == '1'): #Ball is already in backpack
                continue
            
            ballx, bally = ball[i][0], ball[i][1] #balls initial location
            if (ballx == x and bally == y):
                if (sum( map(int, list(backpack))) < c): #have enough capacity
                    new_backpack = copy.deepcopy(backpack)
                    new_backpack = new_backpack[:i] + "1" + new_backpack[i + 1:] 
                    
                    
                    new_state = ((x, y), new_backpack, balls_done)
                    
                    if (not new_state in closedset) and (not (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state) in is_in_heap):
                        par[new_state] = current[1]
                        heappush(openset, (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state))
                        is_in_heap[(cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state)] = 1
                        
        
        if (x + 1 < n):
            if (grid[x + 1][y] != '*'):
                
                new_state = ((x + 1, y), backpack, balls_done)
                if (not new_state in closedset) and (not (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state) in is_in_heap):
                    par[new_state] = current[1]
                    heappush(openset, (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state))
                    is_in_heap[(cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state)] = 1
                    
        if (x - 1 >= 0):
            if (grid[x - 1][y] != '*'):
                new_state = ((x - 1, y), backpack, balls_done)
                
                if (not new_state in closedset) and (not (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state) in is_in_heap):
                    par[new_state] = current[1]
                    heappush(openset, (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state))
                    is_in_heap[(cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state)] = 1
       
        
        if (y + 1 < m):
            if (grid[x][y + 1] != '*'):
                new_state = ((x, y + 1), backpack, balls_done)
                
                if (not new_state in closedset) and (not (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state) in is_in_heap):
                    par[new_state] = current[1]
                    heappush(openset, (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state))
                    is_in_heap[(cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state)] = 1
        
        if (y - 1 >= 0):
            if (grid[x][y - 1] != '*'):
                new_state = ((x, y - 1), backpack, balls_done)
                if (not new_state in closedset) and (not (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state) in is_in_heap):
                    par[new_state] = current[1]
                    heappush(openset, (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state))
                    is_in_heap[(cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state)] = 1
        
st = time.time()
explored = aStar(start_state, goal_state)
fn = time.time()
print("Number of states explored is", explored)
print("Time consumed is %s" %(fn - st))


Start is  ((1, 0), '000', '000')
Path found with len 20
((2, 6), '000', '111')
((2, 5), '100', '011')
((1, 5), '100', '011')
((1, 4), '100', '011')
((1, 3), '100', '011')
((1, 2), '100', '011')
((1, 2), '000', '011')
((1, 1), '001', '010')
((1, 2), '001', '010')
((1, 3), '001', '010')
((2, 3), '001', '010')
((3, 3), '001', '010')
((3, 2), '001', '010')
((3, 2), '000', '010')
((3, 3), '010', '000')
((2, 3), '010', '000')
((2, 3), '000', '000')
((1, 3), '000', '000')
((1, 2), '000', '000')
((1, 1), '000', '000')
((1, 0), '000', '000')
Number of states explored is 211
Time consumed is 0.006096363067626953


### Results of A* with heuristic $h(node) = Euclidean(node.x, node.y, goal.x, goal.y)$:

| --- | Distance with Goal | Satates Seen | Unique States Seen | Avg Exec Time |
| --- | --- | --- | --- | --- |
| Test 1 | 20 | 211 | 211 | 0.003s |
| Test 2 | 48 | 6654 | 6654 | 0.14s |
| Test 3 | 68 | 196789 | 196789 | 4.64s |
| Test 4 | 92 | 802 | 802 | 0.02s |


#### The number of explored states increased since the euclidean heuristic is underestimating the length of the path left in compare to the manhattan heuristic.

# Weighted A*
### We modify the function above by adding a paremeter alpha and choosing the node in $openset$ which has the minimum $g(node) + \alpha\times h(node)$ where $h(node)$ is the manhattan distance of the node with goal state.

In [466]:
from _heapq import heappush, heappop
from math import sqrt

import time

start_state = ( (start_x, start_y), '0' * len(ball), '0' * len(ball) ) #loc, backpack, balls_done
goal_state =  ( (finish_x, finish_y), '0' * len(ball), '1' * len(ball) )
par = dict()

def print_path(cur):
    print(cur)
    if (cur == start_state):
          return
    print_path(par[cur])

def manhattan(x1, y1, x2, y2):
    return abs(x1 - x2) + abs(y1 - y2)
    
def euclidean(x1, y1, x2, y2):
    return sqrt( (x1 - x2)**2 + (y1 - y2)**2 )
    
def get_heuristic(node, goal):
    global alpha
    return alpha * manhattan(node[0][0], node[0][1], goal[0][0], goal[0][1])
    
is_in_heap = dict()
    
def aStar(start, goal):
    
    openset = []
    closedset = set([])
    explored = 0
    current = start
    
    print("Start is ", start)
    
    #openset.append((0, current))
    heappush(openset, (0 + get_heuristic(current, goal), current))
    is_in_heap[(0 + get_heuristic(current, goal), current)] = 1
    
    
    while openset:
    
        current = heappop(openset)
        explored += 1
        del is_in_heap[current]
            
        if current[1] == goal:    
            plen = 0
            node = current[1]
            while (node != start):
                node = par[node]
                plen += 1
            print("Path found with len", plen)
            print_path(goal_state)
            return explored
        
        closedset.add(current[1])
        
        x, y = current[1][0][0], current[1][0][1]
        backpack = current[1][1]
        balls_done = current[1][2]
        cur_gh = current[0]

        
        
        #Put down a ball in no moves
        for i in range(len(backpack)):
            if (backpack[i] == '0'): #dont have it in our backpack
                continue
            if (ball[i][2] == x and ball[i][3] == y):
                backpack = backpack[:i] + "0" + backpack[i + 1:]
                balls_done = balls_done[:i] + "1" + balls_done[i + 1:]
                
                closedset.add(((x, y), backpack, balls_done))
        
        for i in range(len(ball)):
            if (balls_done[i] == '1'): #Ball is already in final destination 
                continue
            if (backpack[i] == '1'): #Ball is already in backpack
                continue
            
            ballx, bally = ball[i][0], ball[i][1] #balls initial location
            if (ballx == x and bally == y):
                if (sum( map(int, list(backpack))) < c): #have enough capacity
                    new_backpack = copy.deepcopy(backpack)
                    new_backpack = new_backpack[:i] + "1" + new_backpack[i + 1:] 
                    
                    
                    new_state = ((x, y), new_backpack, balls_done)
                    
                    if (not new_state in closedset) and (not (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state) in is_in_heap):
                        par[new_state] = current[1]
                        heappush(openset, (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state))
                        is_in_heap[(cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state)] = 1
                        
        
        if (x + 1 < n):
            if (grid[x + 1][y] != '*'):
                
                new_state = ((x + 1, y), backpack, balls_done)
                if (not new_state in closedset) and (not (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state) in is_in_heap):
                    par[new_state] = current[1]
                    heappush(openset, (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state))
                    is_in_heap[(cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state)] = 1
                    
        if (x - 1 >= 0):
            if (grid[x - 1][y] != '*'):
                new_state = ((x - 1, y), backpack, balls_done)
                
                if (not new_state in closedset) and (not (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state) in is_in_heap):
                    par[new_state] = current[1]
                    heappush(openset, (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state))
                    is_in_heap[(cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state)] = 1
       
        
        if (y + 1 < m):
            if (grid[x][y + 1] != '*'):
                new_state = ((x, y + 1), backpack, balls_done)
                
                if (not new_state in closedset) and (not (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state) in is_in_heap):
                    par[new_state] = current[1]
                    heappush(openset, (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state))
                    is_in_heap[(cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state)] = 1
        
        if (y - 1 >= 0):
            if (grid[x][y - 1] != '*'):
                new_state = ((x, y - 1), backpack, balls_done)
                if (not new_state in closedset) and (not (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state) in is_in_heap):
                    par[new_state] = current[1]
                    heappush(openset, (cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state))
                    is_in_heap[(cur_gh + 1 + get_heuristic(new_state, goal) - get_heuristic(current[1], goal), new_state)] = 1
 
alpha = 2
st = time.time()
explored = aStar(start_state, goal_state)
fn = time.time()
print("Number of states explored is", explored)
print("Time consumed is %s" %(fn - st))


Start is  ((1, 0), '000', '000')
Path found with len 32
((2, 6), '000', '111')
((2, 5), '100', '011')
((1, 5), '100', '011')
((1, 4), '100', '011')
((1, 3), '100', '011')
((1, 2), '100', '011')
((1, 2), '000', '011')
((1, 1), '001', '010')
((1, 2), '001', '010')
((1, 3), '001', '010')
((1, 4), '001', '010')
((1, 5), '001', '010')
((2, 5), '001', '010')
((2, 4), '001', '010')
((2, 3), '001', '010')
((3, 3), '001', '010')
((3, 2), '001', '010')
((3, 2), '000', '010')
((3, 3), '010', '000')
((3, 4), '010', '000')
((3, 5), '010', '000')
((2, 5), '010', '000')
((2, 4), '010', '000')
((2, 3), '010', '000')
((2, 3), '000', '000')
((2, 4), '000', '000')
((2, 5), '000', '000')
((1, 5), '000', '000')
((1, 4), '000', '000')
((1, 3), '000', '000')
((1, 2), '000', '000')
((1, 1), '000', '000')
((1, 0), '000', '000')
Number of states explored is 259
Time consumed is 0.007688760757446289


### Results of  Weighted A* with $\alpha = 2$:

| --- | Distance with Goal | Satates Seen | Unique States Seen | Avg Exec Time |
| --- | --- | --- | --- | --- |
| Test 1 | 32 | 259 | 259 | 0.004s |
| Test 2 | 48 | 49 | 49 | 0.004s |
| Test 3 | 268 | 81079 | 81079 | 1.66 |
| Test 4 | 92 | 339 | 339 | 0.015s |

The number of states seen is equal to unique states seen since this algorithm doesn't explore a state more than once.

#### As it can be seen in Test 3, the number of explored states has decreased (81079 < 98885) in compare to non-weighted A* where $\alpha = 1$

### The distance is not optimal anymore.

### Results of  Weighted A* with $\alpha = 5$:

| --- | Distance with Goal | Satates Seen | Unique States Seen | Avg Exec Time |
| --- | --- | --- | --- | --- |
| Test 1 | 38 | 315 | 315 | 0.006s |
| Test 2 | 48 | 49 | 49 | 0.003s |
| Test 3 | 268 | 58939 | 58939 | 1.19 |
| Test 4 | 92 | 351 | 351 | 0.013s |


### It can be seen that the number of states seen has decreased in compare to $\alpha=2$ in tests 2 and 3.


### The distance is not optimal anymore.
